src/Assessment/Processor/ProcessDatasets.cs

```diff
@@ -1243,6 +1249,277 @@ private void UpdateIndividualWebAppPaaSCosts(List<WebApp_PaaS> WebApp_PaaS_List)
+                    }
+                    if (!EnvironmentToUniqueIdMap["Prod"].Contains(sqlinstance.MachineId))
+                        EnvironmentToUniqueIdMap["Prod"].Add(sqlinstance.MachineId);
+                }
+                EnvironmentCountDictionary[sqlinstance.Environment] = +1;
+            }
+
+            SQL_MI_PaaSMachineObj.DevMachinesCountTarget = EnvironmentCountDictionary["Dev"];
+            SQL_MI_PaaSMachineObj.ProdMachinesCountTarget = EnvironmentCountDictionary["Prod"];
+            SQL_MI_PaaSMachineObj.DevUniqueMachinesSource = EnvironmentToUniqueIdMap["Dev"].Count;
+            SQL_MI_PaaSMachineObj.ProdUniqueMachinesSource = EnvironmentToUniqueIdMap["Prod"].Count;
+            return SQL_MI_PaaSMachineObj;
+        }
+
+        private WebApp_IaaS_FS DetailsWebapp_IaaS(List<WebApp_IaaS_Server_Rehost_Perf> WebApp_IaaS_Server_Rehost_Perf_List)
+        {
+            WebApp_IaaS_FS WebappIaaSMachineObj = new WebApp_IaaS_FS();
+            Dictionary<string, int> EnvironmentCountDictionary = new Dictionary<string, int>();
+            Dictionary<string, HashSet<string>> EnvironmentToUniqueIdMap = new Dictionary<string, HashSet<string>>();
+            EnvironmentToUniqueIdMap["Dev"] = new HashSet<string>();
+            EnvironmentToUniqueIdMap["Prod"] = new HashSet<string>();
+
+            foreach (var webapp in WebApp_IaaS_Server_Rehost_Perf_List)
+            {
+                if (webapp.Environment.Equals("Dev"))
+                {
+                    if (!EnvironmentCountDictionary.ContainsKey("Dev"))
+                    {
+                        EnvironmentCountDictionary.Add("Dev", 0);
+                    }
+                    if (!EnvironmentToUniqueIdMap["Dev"].Contains(webapp.MachineId))
+                        EnvironmentToUniqueIdMap["Dev"].Add(webapp.MachineId);
+                }
+                else
+                {
+                    if (!EnvironmentCountDictionary.ContainsKey("Prod"))
+                    {
+                        EnvironmentCountDictionary.Add("Prod", 0);
+                    }
+                    if (!EnvironmentToUniqueIdMap["Prod"].Contains(webapp.MachineId))
+                        EnvironmentToUniqueIdMap["Prod"].Add(webapp.MachineId);
+                }
+                EnvironmentCountDictionary[webapp.Environment] = +1;
+            }
+
+            WebappIaaSMachineObj.DevMachinesCountTarget = EnvironmentCountDictionary["Dev"];
+            WebappIaaSMachineObj.ProdMachinesCountTarget = EnvironmentCountDictionary["Prod"];
+            WebappIaaSMachineObj.DevUniqueMachinesSource = EnvironmentToUniqueIdMap["Dev"].Count;
+            WebappIaaSMachineObj.ProdUniqueMachinesSource = EnvironmentToUniqueIdMap["Prod"].Count;
+            WebappIaaSMachineObj.EnvironmentToUniqueIdMap = EnvironmentToUniqueIdMap;
+            WebappIaaSMachineObj.EnvironmentCountDictionary = EnvironmentCountDictionary;
+            return WebappIaaSMachineObj;
+        }

```

Comment:
P0: there is no need for a separate dictionary for target and source count. Source count == Target count for IaaS_Server_Rehost.
P1: Can this logic be moved to AzureIaaSCalculator to avoid looping on the same dataset again and again.